### PR TITLE
Implement API methods for Block/Fluid states and FluidType [API-8]

### PR DESCRIPTION
**Pull request by i509VCB**
Implements all API methods from Block/FluidState and FluidType.

Some methods may still need to be migrated/adapted from the old BlockState mixin which is now `BlockStateMixin_API_Old` or belong in the data implementation.

### CODE REVIEW

File: src/main/java/org/spongepowered/common/mixin/api/mcp/block/BlockStateMixin_API.java [hunk 1]
```diff
@@ -24,260 +24,44 @@
...
-    default <T extends Property<?, ?>> Optional<T> getProperty(final Direction direction, final Class<T> clazz) {
-        return Optional.empty();
+    public BlockType getType() {
+        return (BlockType) this.shadow$getBlock();
     }
 
-    @Override
-    default List<Immutable<?, ?>> getManipulators() {
-        return Collections.emptyList();
+    @Intrinsic
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/api/mcp/block/BlockStateMixin_API.java]: It’s not intrinsic, the return types are completely different (api.FluidState vs nms.FluidState)

File: src/main/java/org/spongepowered/common/mixin/api/mcp/fluid/FluidMixin_API.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    @Shadow public abstract net.minecraft.state.StateContainer<Fluid, IFluidState> shadow$getStateContainer();
+    @Shadow public abstract net.minecraft.fluid.IFluidState shadow$getDefaultState();
+
+    @Override
+    public ImmutableList<FluidState> getValidStates() {
+        return (ImmutableList) this.shadow$getStateContainer().getValidStates();
+    }
+
+    @Intrinsic
+    public FluidState getDefaultState() {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/api/mcp/fluid/FluidMixin_API.java:52]: Not intrinsic either, it's an API construct, not the `net.minecraft.FluidState`.

File: src/main/java/org/spongepowered/common/mixin/api/mcp/fluid/IFluidStateMixin_API.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        return (FluidType) this.shadow$getFluid();
+    }
+
+    @Override
+    default BlockState getBlock() {
+        return (BlockState) this.getBlockState();
+    }
+
+    @Intrinsic
+    default boolean isEmpty() {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/api/mcp/fluid/IFluidStateMixin_API.java]: If this is intrinsic (which it seems like it very well is in dev time), you need to prefix with a prefix for the
`@Implements(iface = ...)` on the mixing itself as well.

File: src/main/java/org/spongepowered/common/mixin/invalid/api/mcp/block/BlockStateMixin_API_Old.java [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+
+@Mixin(net.minecraft.block.BlockState.class)
+public interface BlockStateMixin_API_Old extends net.minecraft.block.BlockState, BlockState {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/invalid/api/mcp/block/BlockStateMixin_API_Old.java]: What's this `_Old`? is it being retained? Is it still being applied?

File: src/main/java/org/spongepowered/common/mixin/invalid/api/mcp/block/state/StateImplementationMixin_API.java [hunk 5]
```diff
@@ -68,7 +68,7 @@
 import javax.annotation.Nullable;
 
 /**
- * This shares implementation with {@link BlockStateMixin_API}, since this
+ * This shares implementation with {@link BlockStateMixin_API_Old}, since this
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/invalid/api/mcp/block/state/StateImplementationMixin_API.java]: It shouldn't be.

File: src/main/java/org/spongepowered/common/registry/builtin/supplier/ContainerTypeSupplier.java [hunk 6]
```diff
@@ -1,3 +1,27 @@
...
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/registry/builtin/supplier/ContainerTypeSupplier.java]: Unrelated, just undo for now. the licenses can be fixed outside the PR.

File: src/main/java/org/spongepowered/common/mixin/inventory/api/container/ContainerTypeMixin_API.java [hunk 7]
```diff
@@ -1,3 +1,27 @@
...
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/inventory/api/container/ContainerTypeMixin_API.java]: Undo for now, licenses should be fixed outside the PR.

**Review by gabizou — CHANGES_REQUESTED** [review]

File: src/main/java/org/spongepowered/common/mixin/invalid/api/mcp/block/state/StateImplementationMixin_API.java (hunk 5 above)
Comment (i509VCB) [reply src/main/java/org/spongepowered/common/mixin/invalid/api/mcp/block/state/StateImplementationMixin_API.java]: I'd assume the old mixin shouldn't have a shared impl, but the way you've put it makes me think this mixin doesn't really belong here.

File: src/main/java/org/spongepowered/common/mixin/api/mcp/state/StateHolderMixin_API.java [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+import net.minecraft.state.StateHolder;
+import org.spongepowered.api.state.State;
+import org.spongepowered.api.state.StateProperty;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
+
+import java.util.Optional;
+
+@Mixin(StateHolder.class)
+public abstract class StateHolderMixin_API<S extends State<S>, C> implements IStateHolderMixin_API<S> {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/api/mcp/state/StateHolderMixin_API.java:37]: What is `C`? Is this from an implementation's target?

Comment (gabizou) [reply src/main/java/org/spongepowered/common/mixin/api/mcp/state/StateHolderMixin_API.java:37]: This is a case where the class needs javadocs explaining the diamonding of generic types.

**Review by gabizou — APPROVED** [review]

**Pull request merged**